File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/subscriptions/SubscriptionListParams.java
package org.hyperledger.iroha.android.subscriptions;

import java.util.LinkedHashMap;
import java.util.Map;

/** Query parameters for subscription listings. */
public final class SubscriptionListParams {

  private final String ownedBy;
  private final String provider;
  private final SubscriptionStatus status;
  private final Long limit;
  private final Long offset;

  private SubscriptionListParams(final Builder builder) {
    this.ownedBy = normalizeOptional(builder.ownedBy);
    this.provider = normalizeOptional(builder.provider);
    this.status = builder.status;
    this.limit = builder.limit;
    this.offset = builder.offset;
  }

  public String ownedBy() {
    return ownedBy;
  }

  public String provider() {
    return provider;
  }

  public SubscriptionStatus status() {
    return status;
  }

  public Long limit() {
    return limit;
  }

  public Long offset() {
    return offset;
  }

  public Map<String, String> toQueryParameters() {
    final Map<String, String> params = new LinkedHashMap<>();
    if (ownedBy != null) {
      params.put("owned_by", ownedBy);
    }
    if (provider != null) {
      params.put("provider", provider);
    }
    if (status != null) {
      params.put("status", status.slug());
    }
    if (limit != null) {
      params.put("limit", String.valueOf(limit));
    }
    if (offset != null) {
      params.put("offset", String.valueOf(offset));
    }
    return params;
  }

  public static Builder builder() {
    return new Builder();
  }

  public static final class Builder {
    private String ownedBy;
    private String provider;
    private SubscriptionStatus status;
    private Long limit;
    private Long offset;

    private Builder() {}

    public Builder ownedBy(final String ownedBy) {
      this.ownedBy = ownedBy;
      return this;
    }

    public Builder provider(final String provider) {
      this.provider = provider;
      return this;
    }

    public Builder status(final SubscriptionStatus status) {
      this.status = status;
      return this;
    }

    public Builder limit(final Long limit) {
      if (limit != null && limit < 0) {
        throw new IllegalArgumentException("limit must be non-negative");
      }
      this.limit = limit;
      return this;
    }

    public Builder offset(final Long offset) {
      if (offset != null && offset < 0) {
        throw new IllegalArgumentException("offset must be non-negative");
      }
      this.offset = offset;
      return this;
    }

    public SubscriptionListParams build() {
      return new SubscriptionListParams(this);
    }
  }

  private static String normalizeOptional(final String value) {
    if (value == null) {
      return null;
    }
    final String trimmed = value.trim();
    return trimmed.isEmpty() ? null : trimmed;
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/subscriptions/SubscriptionListParams.kt
package org.hyperledger.iroha.sdk.subscriptions

/** Query parameters for subscription listings. */
class SubscriptionListParams(
    ownedBy: String? = null,
    provider: String? = null,
    status: SubscriptionStatus? = null,
    limit: Long? = null,
    offset: Long? = null,
) {
    val ownedBy: String? = normalizeOptional(ownedBy)
    val provider: String? = normalizeOptional(provider)
    val status: SubscriptionStatus? = status

    val limit: Long? = limit?.also {
        require(it >= 0) { "limit must be non-negative" }
    }

    val offset: Long? = offset?.also {
        require(it >= 0) { "offset must be non-negative" }
    }

    fun toQueryParameters(): Map<String, String> = buildMap {
        this@SubscriptionListParams.ownedBy?.let { put("owned_by", it) }
        this@SubscriptionListParams.provider?.let { put("provider", it) }
        status?.let { put("status", it.slug) }
        this@SubscriptionListParams.limit?.let { put("limit", it.toString()) }
        this@SubscriptionListParams.offset?.let { put("offset", it.toString()) }
    }
}

private fun normalizeOptional(value: String?): String? {
    if (value == null) return null
    val trimmed = value.trim()
    return trimmed.ifEmpty { null }
}
